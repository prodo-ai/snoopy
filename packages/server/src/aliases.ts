import * as path from "path";

interface Alias {
  regex: RegExp;
  getNewFilename: (oldFilename: string, parent: string) => string;
}

const aliases: Alias[] = [
  {
    regex: /^styled-components$/,
    getNewFilename: (_: string, parent: string) =>
      path.relative(path.dirname(parent), require.resolve("styled-components")),
  },
];

export default (bundler: any) => {
  const resolver = bundler.resolver;
  const originalResolve = resolver.resolve;

  // Patching
  // https://github.com/parcel-bundler/parcel/blob/master/packages/core/parcel-bundler/src/Resolver.js#L32
  resolver.resolve = async (input: string, parent: string) => {
    let filename = input;

    for (const {regex, getNewFilename} of aliases) {
      if (regex.test(filename)) {
        filename = getNewFilename(filename, parent);
        break;
      }
    }

    return originalResolve.call(resolver, filename, parent);
  };
};
