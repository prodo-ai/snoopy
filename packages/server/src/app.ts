import * as Express from "express";
import * as Bundler from "parcel-bundler";
import * as path from "path";

const clientDir = path.resolve(__dirname, "../../ui");
const entryFile = path.resolve(clientDir, "./public/index.html");

export default (outDir: string, outFile: string) => {
  const app = Express();

  const options = {
    outDir,
    outFile,
    watch: true,
    cache: false,
    minify: false,
    scopeHoist: false,
    hmr: true,
    detailedReport: false,
    bundleNodeModules: true,
  };

  const bundler = new Bundler(entryFile, options);
  bundler.addAssetType(".ts", require.resolve("./component-asset"));

  app.use(bundler.middleware());
  return app;
};
