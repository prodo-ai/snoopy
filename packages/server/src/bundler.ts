import * as Bundler from "parcel-bundler";
import * as path from "path";

export default (
  {
    clientDir,
    outDir,
    outFile,
    searchDir,
    componentsFile,
  }: {
    clientDir: string;
    outDir: string;
    outFile: string;
    searchDir: string;
    componentsFile: string;
  },
  options: any,
) => {
  const entryFile = path.join(clientDir, "public", "index.html");

  process.env.PRODO_SEARCH_DIRECTORY = searchDir;
  process.env.PRODO_COMPONENTS_FILE = path.relative(
    path.join(clientDir, "src", "App"),
    componentsFile,
  );

  options = {
    outDir,
    outFile,
    watch: true,
    cache: false,
    minify: false,
    scopeHoist: false,
    hmr: true,
    detailedReport: false,
    bundleNodeModules: true,
    auotInstall: false,
    ...options,
  };

  const bundler = new Bundler(entryFile, options);
  bundler.addAssetType(".ts", require.resolve("./component-asset"));

  return bundler;
};
