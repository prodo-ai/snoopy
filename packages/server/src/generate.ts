import {searchCodebase} from "@prodo/snoopy-search";
import * as path from "path";

const by = <T extends {}>(transform: (a: T) => string) => (a: T, b: T) =>
  transform(a).localeCompare(transform(b));

export const generateComponentsFileContents = async (
  clientDir: string,
  searchDir: string,
): Promise<string> => {
  const imports = await searchCodebase(searchDir);

  let componentCounter = 0;
  const componentFiles = imports.componentFiles
    .sort(by(f => f.filepath))
    .map(file => ({
      ...file,
      fileExports: file.fileExports
        .sort(by(ex => (ex.isDefaultExport ? "default" : ex.name)))
        .map(ex => ({
          ...ex,
          id: `Component${componentCounter++}`,
        })),
    }));

  let themeCounter = 0;
  const themeFiles = imports.themeFiles.sort(by(f => f.filepath)).map(file => ({
    ...file,
    fileExports: file.fileExports
      .sort(by(ex => (ex.isDefaultExport ? "default" : ex.name)))
      .map(ex => ({
        ...ex,
        id: `Theme${themeCounter++}`,
      })),
  }));

  const importsByFile: {
    [path: string]: {
      defaultImport?: string;
      namedImports: Array<{from: string; to: string}>;
    };
  } = {};

  componentFiles.concat(themeFiles).forEach(({filepath, fileExports}) => {
    if (importsByFile[filepath] == null) {
      importsByFile[filepath] = {
        namedImports: [] as Array<{from: string; to: string}>,
      };
    }
    fileExports.forEach(ex => {
      if (ex.isDefaultExport) {
        if (importsByFile[filepath].defaultImport) {
          throw new Error("Duplicate default import detected.");
        }
        importsByFile[filepath].defaultImport = ex.id;
      } else {
        importsByFile[filepath].namedImports.push({from: ex.name, to: ex.id});
      }
    });
  });

  const importLines = Object.keys(importsByFile)
    .sort()
    .map(filepath => {
      const {defaultImport, namedImports} = importsByFile[filepath];
      return `import ${[
        defaultImport,
        namedImports.length > 0
          ? `{${namedImports
              .map(({from, to}) => `${from} as ${to}`)
              .join(", ")}}`
          : undefined,
      ]
        .filter(x => x != null)
        .join(", ")} from "${path.relative(clientDir, filepath)}";`;
    })
    .join("\n");

  const componentsArrayString = componentFiles
    .map(({filepath, fileExports}) =>
      fileExports
        .map(
          ex =>
            `{path: "${path.relative(searchDir, filepath)}", name: "${
              ex.isDefaultExport ? "default" : ex.name
            }", component: ${ex.id}}`,
        )
        .join(",\n  "),
    )
    .join(",\n  ");

  const themesArrayString = themeFiles
    .map(({filepath, fileExports}) =>
      fileExports
        .map(
          ex =>
            `{path: "${path.relative(searchDir, filepath)}", name: "${
              ex.isDefaultExport ? "default" : ex.name
            }", theme: ${ex.id}}`,
        )
        .join(",\n  "),
    )
    .join(",\n  ");

  return `
${importLines}

import * as React from "react";
import * as ReactDOM from "react-dom";

export { React as UserReact, ReactDOM as UserReactDOM };

export let StyledComponents;
try {
  StyledComponents = require("styled-components");
} catch (e) {
  StyledComponents = null;
}

export const components = [
  ${componentsArrayString}
];

export const themes = [
  ${themesArrayString}
];
`.trimLeft();
};
