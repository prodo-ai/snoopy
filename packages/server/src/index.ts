import {checkMatch} from "@prodo-ai/snoopy-search";
import * as chokidar from "chokidar";
import * as Express from "express";
import * as fs from "fs";
import * as http from "http";
import makeDir = require("make-dir");
import * as path from "path";
import {promisify} from "util";
import applyAliases from "./aliases";
import createBundler from "./bundler";
import registerEndpoints from "./rest";
import registerWebsockets from "./websockets";

const writeFile = promisify(fs.writeFile);

const clientDir = path.dirname(
  path.dirname(require.resolve("@prodo-ai/snoopy-ui")),
);
const outDir = path.resolve(clientDir, "dist");
const outFile = path.resolve(outDir, "index.html");
const defaultPort = 3042;
const startingPort = process.env.PORT
  ? parseInt(process.env.PORT, 10)
  : defaultPort;
const portTriesLimit = 100;

export const start = async (
  port: number = startingPort,
  searchDir = process.cwd(),
) => {
  const app = Express();

  const prodoComponentsModule = path.join(
    searchDir,
    "node_modules",
    "@prodo",
    "components",
  );
  const componentsFile = path.join(prodoComponentsModule, "index.ts");

  await makeDir(path.dirname(componentsFile));
  await writeFile(componentsFile, "");

  // Parcel expects folders in node_modules to have package.json
  const packageFile = path.join(prodoComponentsModule, "package.json");
  await writeFile(
    packageFile,
    JSON.stringify({name: "@prodo/components", main: "index.ts"}),
  );

  const bundler = createBundler({
    clientDir,
    outDir,
    outFile,
    searchDir,
    componentsFile,
  });
  applyAliases(bundler);

  await bundler.bundle();

  app.use(Express.static(outDir));

  chokidar
    .watch(".", {ignored: /(^|[\/\\])\../})
    .on("all", async (_, filename) => {
      const matches = await checkMatch(filename);
      if (matches) {
        // We need to do this to avoid compiling and pushing `filename` at the
        // same time as `componentsFile`.
        await (bundler as any).onChange(componentsFile);
      }
    });

  process.stdout.write(`Starting server on port ${port}...\n`);

  const server = new http.Server(app);
  const ws = registerWebsockets(server);

  registerEndpoints(app, ws, searchDir);

  app.get("/*", (_, response) => {
    response.sendFile((bundler as any).mainBundle.name);
  });

  const listen = (portNumber: number) => {
    app
      .listen(portNumber, () => {
        process.stdout.write(
          `Server is running at http://localhost:${portNumber}.\n`,
        );
      })
      .on("error", e => {
        if (portNumber - startingPort > portTriesLimit) {
          process.stdout.write(`Tried ${portTriesLimit} ports, giving up.`);
        } else if ((e as any).code === "EADDRINUSE") {
          process.stdout.write(
            `Port ${portNumber} is busy, trying ${portNumber + 1}...\n`,
          );
          listen(portNumber + 1);
        } else {
          process.stdout.write(`${e.message}\n`);
        }
      });
  };

  listen(startingPort);
};
