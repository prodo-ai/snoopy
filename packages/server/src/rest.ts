import * as express from "express";
import * as WebSocket from "ws";
import { WebSocketEvents } from "@prodo/snoopy-api";
import * as bodyParser from "body-parser";
import * as path from "path";

export default (app: express.Application, ws: WebSocket.Server, searchDir: string) => {
  app.use(bodyParser.text());
  app.post("/open-file", (req, res) => {
    const file = req.body;
    ws.clients.forEach(client => {
      client.send(JSON.stringify({ type: WebSocketEvents.OPEN_FILE, file: path.relative(searchDir, file) }))
    })
    res.send();
  });
};
