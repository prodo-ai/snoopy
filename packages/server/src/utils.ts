import * as fs from "fs";
import {promisify} from "util";

export const writeFile = promisify(fs.writeFile);
export const stat = promisify(fs.stat);

export const exists = async (filePath: string): Promise<boolean> => {
  try {
    await stat(filePath);
    return true;
  } catch (error) {
    if (error.code === "ENOENT") {
      return false;
    }
    throw error;
  }
};
