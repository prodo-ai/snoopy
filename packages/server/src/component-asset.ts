import * as path from "path";
import {generateComponentsFileContents, generateLibs} from "./generate";

// tslint:disable-next-line:no-submodule-imports
import TypeScriptAsset = require("parcel-bundler/src/assets/TypeScriptAsset");

const componentsFileRegex = /@prodo\/components\/index\.ts$/;
const libsFileRegex = /@prodo\/components\/lib.ts$/;

class ComponentAsset extends TypeScriptAsset {
  public async load() {
    if (componentsFileRegex.test(this.name)) {
      const fileDir = path.dirname(this.name);
      this.contents = await generateComponentsFileContents(
        fileDir,
        this.options.env.PRODO_SEARCH_DIRECTORY,
      );

      return this.contents;
    } else if (libsFileRegex.test(this.name)) {
      this.contents = generateLibs();
      return this.contents;
    }

    return super.load();
  }
}

module.exports = ComponentAsset;
