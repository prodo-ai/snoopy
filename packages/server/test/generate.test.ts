import * as path from "path";
import {generateComponentsFileContents} from "../src/generate";

const searchDir = path.resolve(__dirname, "fixtures", "example");
const brokenDir = path.resolve(__dirname, "fixtures", "broken");
const clientDir = path.resolve(__dirname, "fixtures");

describe("generateComponentsFileContents", () => {
  it("detects inside javascript", async () => {
    const contents = await generateComponentsFileContents(clientDir, searchDir);
    expect(
      /import {Button as Component\d+} from "example\/javascript.jsx";/.test(
        contents,
      ),
    ).toBe(true);
  });
  it("detects inside typescript", async () => {
    const contents = await generateComponentsFileContents(clientDir, searchDir);
    expect(
      /import {Button as Component\d+} from "example\/typescript\.tsx";/.test(
        contents,
      ),
    ).toBe(true);
  });
  it("detects inside index files", async () => {
    const contents = await generateComponentsFileContents(clientDir, searchDir);
    expect(
      /import {Button as Component\d+} from "example";/.test(contents),
    ).toBe(true);
  });

  it("finds named component exports", async () => {
    const contents = await generateComponentsFileContents(clientDir, searchDir);
    expect(
      /import {Button as Component\d+} from "example\/components\/NamedExport\.tsx";/.test(
        contents,
      ),
    ).toBe(true);
  });

  it("finds default component exports", async () => {
    const contents = await generateComponentsFileContents(clientDir, searchDir);
    expect(
      /import Component\d+ from "example\/components\/DefaultExport\.tsx";/.test(
        contents,
      ),
    ).toBe(true);
  });

  it("finds mixed component exports", async () => {
    const contents = await generateComponentsFileContents(clientDir, searchDir);
    expect(
      /import Component\d+, {Button as Component\d+, Button2 as Component\d+} from "example\/components\/MultipleExports\.tsx";/.test(
        contents,
      ),
    ).toBe(true);
  });

  it("ignores component files with no marked exports", async () => {
    const contents = await generateComponentsFileContents(clientDir, searchDir);
    expect(
      /import .* from "example\/components\/Unmarked\.tsx";/.test(contents),
    ).toBe(false);
  });

  it("ignores unmarked component exports", async () => {
    const contents = await generateComponentsFileContents(clientDir, searchDir);
    expect(
      /import {Button2 as Component\d+} from "example\/components\/Mixed\.tsx";/.test(
        contents,
      ),
    ).toBe(true);
  });

  it("finds named theme exports", async () => {
    const contents = await generateComponentsFileContents(clientDir, searchDir);
    expect(
      /import {theme as Theme\d+} from "example\/themes\/named-export\.ts";/.test(
        contents,
      ),
    ).toBe(true);
  });

  it("finds default theme exports", async () => {
    const contents = await generateComponentsFileContents(clientDir, searchDir);
    expect(
      /import Theme\d+ from "example\/themes\/default-export\.ts";/.test(
        contents,
      ),
    ).toBe(true);
  });

  it("finds mixed theme exports", async () => {
    const contents = await generateComponentsFileContents(clientDir, searchDir);
    expect(
      /import Theme\d+, {theme as Theme\d+, theme2 as Theme\d+} from "example\/themes\/multiple-exports\.ts";/.test(
        contents,
      ),
    ).toBe(true);
  });

  it("ignores theme files with no marked exports", async () => {
    const contents = await generateComponentsFileContents(clientDir, searchDir);
    expect(
      /import .* from "example\/themes\/unmarked\.ts";/.test(contents),
    ).toBe(false);
  });

  it("ignores unmarked theme exports", async () => {
    const contents = await generateComponentsFileContents(clientDir, searchDir);
    expect(
      /import {theme2 as Theme\d+} from "example\/themes\/mixed\.ts";/.test(
        contents,
      ),
    ).toBe(true);
  });

  it("finds components and themes in the same file", async () => {
    const contents = await generateComponentsFileContents(clientDir, searchDir);
    expect(
      /import {Button as Component\d+, theme as Theme\d+} from "example\/mixed\.tsx";/.test(
        contents,
      ),
    ).toBe(true);
  });

  // Currently failing
  it.skip("doesn't fail on directories with broken files", async () => {
    const contents = await generateComponentsFileContents(clientDir, brokenDir);

    expect(
      /import {Button as Component\d+, theme as Theme\d+} from "good\.tsx";/.test(
        contents,
      ),
    ).toBe(true);
    expect(/import .* from "examples\/broken\.ts";/.test(contents)).toBe(false);
  });

  it("snapshot test", async () => {
    const contents = await generateComponentsFileContents(clientDir, searchDir);

    expect(contents).toMatchSnapshot();
  });
});
