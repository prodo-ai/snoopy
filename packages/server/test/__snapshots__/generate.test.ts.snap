// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateComponentsFileContents snapshot test 1`] = `
"import Component0 from \\"example/components/DefaultExport.tsx\\";
import {Button2 as Component1} from \\"example/components/Mixed.tsx\\";
import Component4, {Button as Component2, Button2 as Component3} from \\"example/components/MultipleExports.tsx\\";
import {Button as Component5} from \\"example/components/NamedExport.tsx\\";
import {Button as Component6} from \\"example/index.jsx\\";
import {Button as Component7} from \\"example/javascript.jsx\\";
import {Button as Component8, theme as Theme0} from \\"example/mixed.tsx\\";
import Theme1 from \\"example/themes/default-export.ts\\";
import {theme2 as Theme2} from \\"example/themes/mixed.ts\\";
import Theme5, {theme as Theme3, theme2 as Theme4} from \\"example/themes/multiple-exports.ts\\";
import {theme as Theme6} from \\"example/themes/named-export.ts\\";
import {Button as Component9} from \\"example/typescript.tsx\\";

import * as React from \\"react\\";
import * as ReactDOM from \\"react-dom\\";

export { React as UserReact, ReactDOM as UserReactDOM };

export let StyledComponents;
try {
  StyledComponents = require(\\"styled-components\\");
} catch (e) {
  StyledComponents = null;
}

export const components = [
  {path: \\"components/DefaultExport.tsx\\", name: \\"default\\", component: Component0},
  {path: \\"components/Mixed.tsx\\", name: \\"Button2\\", component: Component1},
  {path: \\"components/MultipleExports.tsx\\", name: \\"Button\\", component: Component2},
  {path: \\"components/MultipleExports.tsx\\", name: \\"Button2\\", component: Component3},
  {path: \\"components/MultipleExports.tsx\\", name: \\"default\\", component: Component4},
  {path: \\"components/NamedExport.tsx\\", name: \\"Button\\", component: Component5},
  {path: \\"index.jsx\\", name: \\"Button\\", component: Component6},
  {path: \\"javascript.jsx\\", name: \\"Button\\", component: Component7},
  {path: \\"mixed.tsx\\", name: \\"Button\\", component: Component8},
  {path: \\"typescript.tsx\\", name: \\"Button\\", component: Component9}
];

export const themes = [
  {path: \\"mixed.tsx\\", name: \\"theme\\", theme: Theme0},
  {path: \\"themes/default-export.ts\\", name: \\"default\\", theme: Theme1},
  {path: \\"themes/mixed.ts\\", name: \\"theme2\\", theme: Theme2},
  {path: \\"themes/multiple-exports.ts\\", name: \\"theme\\", theme: Theme3},
  {path: \\"themes/multiple-exports.ts\\", name: \\"theme2\\", theme: Theme4},
  {path: \\"themes/multiple-exports.ts\\", name: \\"default\\", theme: Theme5},
  {path: \\"themes/named-export.ts\\", name: \\"theme\\", theme: Theme6}
];
"
`;
